RFC: improve inlined line numbers, take 2

JeffBezanson opened the pull request:
This is a rebased and edited version of #14949. To reduce the scope a bit, I changed it to store file names and function names instead of references to LambdaInfos. The goal is to get something mergeable as soon as possible, as this change has several important fixes from @carnaval :
1. It fixes the formatting of our LLVM debug info and represents inlined functions properly.
2. It changes the representation of source location info in our IR.
3. It adds the ability to represent any number of nested inlined functions in backtraces.

I want to get these changes in even if the backtraces are not perfect yet. We can improve it later, either by adding back the LambdaInfo (or Method) references or by keeping a mapping from filename and line number to Methods.

I will merge #15583 into this branch, so we can take advantage of some of this for macro code locations as well.

carnaval committed:
Improve inlined line numbers

tkelman commented on src/debuginfo.cpp:
```diff
@@ -759,6 +769,10 @@ extern "C" void jl_register_fptrs(uint64_t sysimage_base, void **fptrs, jl_lambd
 }
 
+#ifdef _OS_LINUX_
+#include <link.h>
```
top of file, around L53-54?

JeffBezanson committed:
keep track of filenames in IR with metadata push/pop

- add location info to empty blocks; fixes #15280
- remove file names from non-initial `line` nodes
- emit `(meta push_loc filename)` and `(meta pop_loc)` around nested
  blocks from different files, to handle macros
- fix line number of `catch` blocks

JeffBezanson committed:
report un-named "inlined" code as being from a macro expansion

JeffBezanson commented:
I'm not sure how to handle locations from macro expansions. My guess is that there is not good support for the filename magically switching in the middle of a function, so for now I handled it the same way as an inlined function (since that seems to work well), with the inlined function name being `macro expansion`.

toivoh commented:
Will there be a way for Julia code examining the AST to figure or which file that lines from a macro expansion came from?

StefanKarpinski commented:
> I'm not sure how to handle locations from macro expansions.

I think ultimately the only way to handle generated code sanely will be to keep generated ASTs and be able to step through them. The `julia` program should have a command-line argument `--keep-source=[yes|no]` which defaults to "yes" in interactive mode and defaults to "no" in non-interactive mode.

JeffBezanson commented:
> Will there be a way for Julia code examining the AST to figure or which file that lines from a macro expansion came from?

Yes, there are `push_loc` and `pop_loc` for blocks from a different file.

timholy commented:
Big fan of this effort.

> there are push_loc and pop_loc for blocks from a different file.

Awesome! Doesn't yet seem documented in this PR, though. Might need a small section in devdocs or something.

Keno commented on src/codegen.cpp:
```diff
@@ -4286,7 +4277,7 @@ static std::unique_ptr<Module> emit_function(jl_lambda_info_t *lam, jl_llvm_func
         SP = dbuilder.createFunction((DIDescriptor)dbuilder.getCU(),
         #else
-        SP = dbuilder.createFunction(CU,
+        SP = dbuilder.createFunction(topfile,
```
I'm not sure this works. If the function isn't associated to a compile unit, it might not end up being represented in the binary.

Keno commented on src/codegen.cpp:
```diff
@@ -4675,84 +4668,64 @@ static std::unique_ptr<Module> emit_function(jl_lambda_info_t *lam, jl_llvm_func
+                                             inl_name,
+                                             new_file,
+                                             0,
+                                             jl_di_func_sig,
```
This isn't really correct. NOt that anything currently depends on it, but we might want to revisit this at a later point.

Keno commented on src/debuginfo.cpp:
```diff
@@ -814,9 +825,23 @@ bool jl_dylib_DI_for_fptr(size_t pointer, const llvm::object::ObjectFile **obj,
+    int dladdr_success;
+    uint64_t fbase;
+#ifdef _OS_LINUX_
+    // dlinfo.dli_fbase is not the right value for the main executable on linux
```
Should this comment go below?

Keno commented on src/debuginfo.cpp:
```diff
@@ -1228,34 +1249,83 @@ void jl_getFunctionInfo(char **name, char **filename, size_t *line,
+#ifdef _OS_DARWIN_
+#define MAP_ANONYMOUS MAP_ANON
+#endif
+uint8_t *juliaAllocateSection(uintptr_t size, unsigned align, section_type typ)
```
This shouldn't be part of this PR.

JeffBezanson commented:
Dev docs added and comments addressed. Should be ready to merge.

tkelman commented on test/backtrace.jl:
```diff
@@ -4,68 +4,15 @@ bt = backtrace()
 
 @test have_backtrace
 
-# Test location information for inlined code (ref issues #1334 #12544)
```
if the results of these all changed, can we change them to test the affirmative of what they should be now, instead of deleting them altogether?

JeffBezanson replied:
Ah yes, will do.

tkelman commented on src/task.c:
```diff
@@ -617,13 +621,19 @@ void jl_init_tasks(void)
+                                           jl_any_type, jl_any_type,
+                                           jl_any_type, jl_any_type,
+                                           jl_any_type, jl_any_type,
+                                           jl_tupletype_fill(sizeof(jl_jmp_buf), jl_uint8_type),
```
needs a cast to `(jl_value_t *) jl_uint8_type`

JeffBezanson committed:
add back backtrace tests for inlined functions

tkelman commented:
This breaks building against LLVM 3.3, if we're still concerned with trying to keep that working for testing/perf comparison.
https://gist.github.com/02d8c53f6c9e5962cb738292478646bb

carnaval commented:
Thanks a lot. I got sidetracked with the memory issues and was starting to lose faith ;). Glad to see a lot of it was salvageable. Allocating our own jit sections was an easy win if we don't care too much about W^X but I'm not sure that'll be the consensus.

JeffBezanson commented:
Yeah the allocation part is definitely important to work on, but should just be done separately. I suspect it's probably ok to use RWX pages by default, with a switch to disable them.

JeffBezanson commented:
It would make me happy to remove all the LLVM version ifdefs. If we have ifdefs for some LLVM version, but it's not tested in CI and even most developers never run it, is it *really* supported? The readability and maintainability gains would be significant.

However currently I'm having trouble building with LLVM 3.3. I get errors about `std::unique_ptr` and such. There must be another switch I need to flip.

tkelman commented:
`CXXFLAGS=-std=c++11`

I agree that it would help clean things up, but 3.7 is still a pretty bad compiler perf regression relative to 3.3 isn't it? Though I see a failure at the end of bootstrap that was introduced by one of the codegen rewrites in March, so maybe it's already broken and not salvagable.

yuyichao commented:
You need `CXXFLAGS+=' -std=c++0x'` to compile with llvm 3.3 with gcc. I've also seen issue about linker confused by '@' in the symbol table though.

JeffBezanson commented:
Thanks.

> 3.7 is still a pretty bad compiler perf regression relative to 3.3 isn't it?

Yes it is. But we already decided to make that tradeoff.

tkelman commented:
On master, as a default build configuration. We don't have every-commit CI for lots of nominally supported configurations, ARM, FreeBSD, etc. We haven't deleted all of the LLVM 3.3 code yet, so the decision of whether or not to keep that a possible option for testing hasn't totally been finalized.

yuyichao commented:
I believe @vtjnash is testing it on Mac, which doesn't require much special configuration due to clang being permissive about c++11 features and `@` not being a special character for the linker. (FWIW g++ 6.1 use gnu++14 by default so the `CXXFLAGS` setting shouldn't be necessary anymore)

JeffBezanson commented:
OK, LLVM 3.3 build restored to at least as working as it was before.

vtjnash commented:
Is this the cause of the new failure in genstdlib?
```
WARNING: missing docs for signature:

    lookup(pointer::Union{Ptr{Void}, UInt}) -> StackFrame
```

tkelman commented:
81d2c464f0f8e1390e61de0a9ffc87f8b435314b
